File: webapp/views/methodology_view.py
"""Просмотр действующей методики: компоненты, веса, балльные шкалы,
версии, пороги категорий (раздел 4 ТЗ)."""

from __future__ import annotations

from datetime import date

import pandas as pd
import streamlit as st

from webapp.common import get_methodology_repo

st.title("📐 Методика расчёта Индекса")

repo = get_methodology_repo()
versions = repo.all_versions()

version_labels = {m.version: f"{m.version} (действует с {m.effective_date})" for m in versions}
selected_version = st.selectbox(
    "Версия методики", options=list(version_labels.keys()), format_func=lambda v: version_labels[v],
    index=len(versions) - 1,
)
methodology = repo.get_by_version(selected_version)

st.markdown(f"### {methodology.title}")
st.write(
    f"**Версия:** {methodology.version} · **Действует с:** {methodology.effective_date} · "
    f"**Диапазон значений ИУ:** {methodology.value_range[0]:g}–{methodology.value_range[1]:g}"
)
if methodology.approved_by:
    st.write(f"**Утверждена:** {methodology.approved_by}")
else:
    st.warning("Реквизиты утверждения методики не заполнены в конфигурации.")

st.markdown(
    "**ИУ = 0,20·GEO + 0,30·ACC + 0,10·DEM + 0,10·SERV + 0,10·INFRA + 0,10·ECO + 0,10·ECON**"
)

weight_df = pd.DataFrame(
    [{"Компонент": k, "Название": c.title, "Вес в ИУ": c.weight} for k, c in methodology.components.items()]
)
st.bar_chart(weight_df.set_index("Компонент")["Вес в ИУ"])

for component_key, component_spec in methodology.components.items():
    with st.expander(f"{component_key} — {component_spec.title} (вес {component_spec.weight:g})"):
        if component_spec.notes:
            st.caption(component_spec.notes)
        for indicator_key, indicator_spec in component_spec.indicators.items():
            st.markdown(
                f"**{indicator_spec.title}** (вес {indicator_spec.weight:g} внутри {component_key})"
                + (f" — единица: {indicator_spec.unit}" if indicator_spec.unit else "")
            )
            if indicator_spec.notes:
                st.caption(indicator_spec.notes)
            if indicator_spec.type == "numeric_thresholds":
                scale_rows = [
                    {"Интервал": rule.get("label", ""), "Балл": rule["score"]}
                    for rule in indicator_spec.rules or []
                ]
            else:
                scale_rows = [
                    {"Интервал": label, "Балл": score}
                    for _, label, score in indicator_spec.option_choices()
                ]
            st.table(pd.DataFrame(scale_rows))

st.divider()
st.subheader("Категоризация населённых пунктов по уровню уязвимости")
st.write(f"Пороги утверждены: {'да' if methodology.categories_approved() else '⚠️ нет — временная заглушка'}")
threshold_rows = []
for rule in methodology.category_thresholds():
    parts = []
    if "gt" in rule:
        parts.append(f"> {rule['gt']}")
    if "gte" in rule:
        parts.append(f">= {rule['gte']}")
    if "lt" in rule:
        parts.append(f"< {rule['lt']}")
    if "lte" in rule:
        parts.append(f"<= {rule['lte']}")
    threshold_rows.append({"Условие по ИУ": " и ".join(parts), "Категория": rule["category"]})
st.table(pd.DataFrame(threshold_rows))

st.divider()
st.subheader("Пересчёт по дате: какая версия действовала")
picked_date = st.date_input("Дата", value=date.today())
try:
    applicable = repo.get_for_date(picked_date)
    st.success(f"На {picked_date} действовала версия методики **{applicable.version}** "
               f"(с {applicable.effective_date}).")
except Exception as exc:  # noqa: BLE001
    st.error(str(exc))

File: netlify-site/js/views/methodologyView.js
/**
 * Страница «Методика» — аналог `webapp/views/methodology_view.py`.
 */

window.Views = window.Views || {};

window.Views.methodology = function methodologyView(root) {
  const { esc } = window.RenderUtils;
  const methodology = AppData.methodology;

  function render() {
    root.innerHTML = `
      <h1>📐 Методика расчёта Индекса</h1>
      <h3>${esc(methodology.title)}</h3>
      <p><b>Версия:</b> ${esc(methodology.version)} · <b>Действует с:</b> ${esc(methodology.effectiveDate)} ·
        <b>Диапазон значений ИУ:</b> ${methodology.valueRange[0]}–${methodology.valueRange[1]}</p>
      ${methodology.approvedBy ? `<p><b>Утверждена:</b> ${esc(methodology.approvedBy)}</p>`
        : '<div class="alert alert-warning">Реквизиты утверждения методики не заполнены в конфигурации.</div>'}
      <p><b>ИУ = 0,20·GEO + 0,30·ACC + 0,10·DEM + 0,10·SERV + 0,10·INFRA + 0,10·ECO + 0,10·ECON</b></p>
      <canvas id="weights-chart" height="140"></canvas>
      <div id="components-block"></div>
      <hr/>
      <h2>Категоризация населённых пунктов по уровню уязвимости</h2>
      <p>Пороги утверждены: ${methodology.categoriesApproved() ? 'да' : '⚠️ нет — временная заглушка'}</p>
      <table class="data-table">
        <thead><tr><th>Условие по ИУ</th><th>Категория</th></tr></thead>
        <tbody>${(methodology.categories.thresholds || []).map((rule) => `<tr>
          <td>${describeRule(rule)}</td><td>${esc(rule.category)}</td>
        </tr>`).join('')}</tbody>
      </table>
    `;

    renderWeightsChart();
    renderComponents();
  }

  function describeRule(rule) {
    const parts = [];
    if ('gt' in rule) parts.push(`&gt; ${rule.gt}`);
    if ('gte' in rule) parts.push(`&gt;= ${rule.gte}`);
    if ('lt' in rule) parts.push(`&lt; ${rule.lt}`);
    if ('lte' in rule) parts.push(`&lt;= ${rule.lte}`);
    return parts.join(' и ');
  }

  function renderWeightsChart() {
    const canvas = document.getElementById('weights-chart');
    const keys = methodology.componentKeys();
    new Chart(canvas, {
      type: 'bar',
      data: { labels: keys, datasets: [{ data: keys.map((k) => methodology.components[k].weight), backgroundColor: '#2b6cb0' }] },
      options: { plugins: { legend: { display: false } } },
    });
  }

  function renderComponents() {
    const container = document.getElementById('components-block');
    container.innerHTML = methodology.componentKeys().map((compKey) => {
      const comp = methodology.components[compKey];
      const indRows = Object.entries(comp.indicators).map(([indKey, ind]) => {
        let scaleRows;
        if (ind.type === 'numeric_thresholds') {
          scaleRows = (ind.rules || []).map((r) => `<tr><td>${esc(r.label)}</td><td>${r.score}</td></tr>`).join('');
        } else {
          scaleRows = Object.values(ind.options || {}).map((o) => `<tr><td>${esc(o.label)}</td><td>${o.score}</td></tr>`).join('');
        }
        return `<p><b>${esc(ind.title)}</b> (вес ${ind.weight} внутри ${compKey})${ind.unit ? ` — единица: ${esc(ind.unit)}` : ''}</p>
          ${ind.notes ? `<p class="caption">${esc(ind.notes)}</p>` : ''}
          <table class="data-table"><thead><tr><th>Интервал</th><th>Балл</th></tr></thead><tbody>${scaleRows}</tbody></table>`;
      }).join('');
      return `<details class="component-block">
        <summary>${compKey} — ${esc(comp.title)} (вес ${comp.weight})</summary>
        ${comp.notes ? `<p class="caption">${esc(comp.notes)}</p>` : ''}
        ${indRows}
      </details>`;
    }).join('');
  }

  render();
};
